app/src/main/java/org/mozilla/focus/home/NavigationOverlayFragment.kt

```diff
@@ -0,0 +1,122 @@
+ * - Initial homescreen: navigation is placed below the app bar
+ * - Dialog: navigation is full screen, overlaid on a semi opaque overlay
+ *
+ * Note: this was originally implemented as a DialogFragment to better decouple the UI of the overlay from the
+ * app but there was a long delay before it could be displayed so we switched to a regular Fragment.
+ */
+class NavigationOverlayFragment : Fragment() {
+
+    private val isInitialHomescreen: Boolean by lazy { arguments!!.getBoolean(KEY_IS_INITIAL_HOMESCREEN) }
+
+    private val callbacks: BrowserFragmentCallbacks? get() = activity as BrowserFragmentCallbacks?
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        val overlay = inflater.inflate(R.layout.fragment_navigation_overlay, container, false)
+
+        with(overlay.semiOpaqueBackground) {
+            visibility = if (isInitialHomescreen) View.GONE else View.VISIBLE
+            setOnClickListener {
+                dismiss()
+                TelemetryWrapper.dismissHomeOverlayClickEvent()
+            }
+        }
+
+        with(overlay.homeTiles) {
+            onTileClicked = { callbacks?.onNonTextInputUrlEntered(it) }
+            urlSearcher = activity as UrlSearcher
+
+            homeTileLongClickListener = object : HomeTileLongClickListener {
+                override fun onHomeTileLongClick(unpinTile: () -> Unit) {
+                    callbacks?.onHomeTileLongClick(unpinTile)
+                }
+            }
+        }
+
+        setOverlayHeight(overlay.homeTiles)
+
+        return overlay
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+
+        // Since we start the async jobs in View.init and Android is inflating the view for us,
+        // there's no good way to pass in the uiLifecycleJob. We could consider other solutions
+        // but it'll add complexity that I don't think is probably worth it.
+        homeTiles.uiLifecycleCancelJob.cancel(CancellationException("Parent lifecycle has ended"))
+    }
+
+    private fun setOverlayHeight(homeTiles: HomeTileGridNavigation) {
+        homeTiles.updateLayoutParams {
+            // Since we're attached to a full screen view decoupled from the app bar, we need to use
+            // the app bar height to position the navigation below the app bar.
+            //
+            // The current layout implementation makes setting the height through margins convenient.
+            val params = it as ViewGroup.MarginLayoutParams
+            params.topMargin = resources.getDimensionPixelSize(R.dimen.appbar_height)
+        }
+    }
+
+    fun show(fragmentManager: FragmentManager, activity: Activity) {
+        activity.getNavigationOverlayContainer().visibility = View.VISIBLE
+        fragmentManager.beginTransaction()
+            .replace(R.id.navigationOverlayContainer, this, FRAGMENT_TAG)
+            .commit()
```

Comment:
Were we doing this before?  It seems strange for a fragment to control its own attach / detatch logic.

Reply:
No: before this wasn't a fragment nor was it encapsulated so the visibility was being changed by the BrowserFragment.

I have the MainActivity handle this with FragmentLifecycleCallabcks in a later commit:
```
+        private fun setNavigationOverlayContainerIsVisible(isVisible: Boolean) {
+            // We set the visibility with fragment lifecycle callbacks to ensure the container visibility reflects the
+            // fragment view hierarchy state. This seamlessly handles cases where the fragments are modified but we
+            // didn't initiate it (e.g. during state restoration, active fragments will be restored but not the view hierarchy).
+            navigationOverlayContainer.visibility = if (isVisible) View.VISIBLE else View.GONE
+        }
```

Reply:
Ah, I misread this. You reference attach/detach rather than controlling the container view's visibility.

I chose to do it this way because that's how `DialogFragment` from the android framework does it and it kept BrowserFragment cleaner but now I recognize that DialogFragment doesn't attach itself to the existing view hierarchy so it's completely decoupled from the app, unlike the code here. I'll change it.

Reply:
> I'll change it.

One problem I see now is that `dismiss` is called from inside the fragment (clicking the overlay dismiss the fragment) and communicating back to the BrowserFragment to dismiss us (to centralize the code attach/detach code) is 1) tightly coupling and 2) a pain in the ass. I'll leave `dismiss` as is with a comment but this should be much easier when we move to an MVVM model.

Reply:
Then again, MainActivity owns the view so maybe all the logic can be moved there and the code isn't that messy... 🤔

Reply:
Oh neat, this uncovered a bug: when we dismiss from inside the fragment, we don't call the necessary methods BrowserFragment calls when it dismisses. Looking for a solution...

Reply:
I filed https://github.com/mozilla-mobile/firefox-echo-show/issues/167